src/Time/TimeSteppers/RungeKutta4.hpp

```diff
@@ -0,0 +1,218 @@
+                                 const RungeKutta4& /*rhs*/) noexcept {
+  return true;
+}
+
+inline bool constexpr operator!=(const RungeKutta4& /*lhs*/,
+                                 const RungeKutta4& /*rhs*/) noexcept {
+  return false;
+}
+
+template <typename Vars, typename DerivVars>
+void RungeKutta4::update_u(
+    const gsl::not_null<Vars*> u,
+    const gsl::not_null<History<Vars, DerivVars>*> history,
+    const TimeDelta& time_step) const noexcept {
+  const size_t substep = history->size() - 1;
+  const auto& dt_vars = (history->end() - 1).derivative();
+  const auto& U0 = history->begin().value();
+
+  switch (substep) {
+    case 0: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(1) = u^n + dt*RHS(u^n,t^n)/2
+      // On entry V = u^n, U0 = u^n, rhs0 = RHS(u^n, t^n),
+      // time = t^n
+      *u += 0.5 * (time_step.value() * dt_vars);
+      // On exit v = v^(1), time = t^n + (1/2)*dt
+      break;
+    }
+    case 1: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(2) = u^n + dt * RHS(v^(1), t^n + (1/2)*dt)/2
+      // On entry V = v^(1), U0 = u^n, rhs0 = RHS(v^(1), t^n + dt/2),
+      // time = t^n + dt
+      *u = U0 + 0.5 * (time_step.value() * dt_vars);
+      // On exit v = v^(2), time = t^n + (1/2)*dt
+      break;
+    }
+    case 2: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(3) = u^n + dt*RHS(v^(2), t^n + (1/2)*dt))
+      // On entry V = v^(2), U0 = u^n, rhs0 = RHS(v^(2), t^n + (1/2)*dt),
+      // time = t^n + (1/2)*dt
+      *u = U0 + time_step.value() * dt_vars;
+      // On exit v = v^(3), time = t^n + dt
+      break;
+    }
+    case 3: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // u^(n+1) = (2v^(1) + 4*v^(2) + 2*v^(3) + v^(4) - 3*U0)/6
+      // On entry V = v^(3), U0 = u^n, rhs0 = RHS(v^(3), t^n + dt),
+      // time = t^n + dt
+      // Note: v^(4) = U0 + dt*RHS(t+dt, v^(3)); inserting this gives
+      // u^(n+1) = (2v^(1) + 4*v^(2) + 2*v^(3) + dt*RHS(t+dt,v^(3)) - 2*U0)/6
+      constexpr double one_sixth = 1.0 / 6.0;
+      *u = (2.0 * (history->begin() + 1).value() +
+            4.0 * (history->begin() + 2).value() +
+            2.0 * (history->begin() + 3).value() +
+            (time_step.value() * dt_vars - 2.0 * U0)) *
+           one_sixth;
+      // On exit v = u^(n+1), time = t^n + dt
+      break;
+    }
+    default:
+      ERROR("Bad substep value in RK4: " << substep);
+  }
+
+  // Clean up old history
+  if (history->size() == number_of_substeps()) {
+    history->mark_unneeded(history->end());
+  }
+}
+
+template <typename Vars, typename DerivVars>
+void RungeKutta4::dense_update_u(const gsl::not_null<Vars*> u,
+                                 const History<Vars, DerivVars>& history,
+                                 const double time) const noexcept {
+  ASSERT(history.size() == number_of_substeps(),
+         "Can only dense output on last substep");
```

Comment:
Can you give the current substep in the error message and the total number of substeps?